stdlib/Logging/src/ConsoleLogger.jl

```diff
@@ -116,7 +116,13 @@ function handle_message(logger::ConsoleLogger, level::LogLevel, message, _module
 
     # Generate a text representation of the message and all key value pairs,
     # split into lines.
-    msglines = [(indent=0, msg=l) for l in split(chomp(convert(String, string(message))::String), '\n')]
+    msglines = if message isa Base.AnnotatedString
```

Comment:
Feels a bit like a design failure that any custom string type needs to be hard coded here, but maybe it is unavoidable.

Reply:
The specialisation on `String` not `AbstractString` is largely unnoticeable, it just so happens that with an `AnnotatedString` (and only `AnnotatedString`, as far as I'm aware), there's a potential difference in the externally visible behaviour.

Reply:
@fredrikekre do please let me know if you've got any further comments here, otherwise I'll just mark this as resolved shortly.

Reply:
I agree this seems odd. The assert and convert here was added by #38292 and #44500 in the latency hunts; without them there'd be no need for a separate codepath.

But I suppose going back to the simple `[(indent=0, msg=l) for l in split(chomp(string(message)), '\n')]` would likely reintroduce invalidations.  Maybe worth a check?

Reply:
Maybe? I'm not sure I'd be a good person to check however — perhaps someone more familiar with invalidations could take a look if needed?

If this is indeed a little tricky to work out, might it be possible to merge this as an initial 1 -> 2 `AbstractString` widening of the allowed types here, and then explore making this fully generic again in a subsequent PR?

Reply:
Ok, I'm not entirely sure of this, but I think the following might suggest that a widening to "any `AbstractString`" may be premature given Tim's invalidation concerns.

```julia
# What it used to be before Tim's PRs
julia> m0(message) = [(indent=0, msg=l) for l in split(chomp(string(message)), '\n')]
m0 (generic function with 1 method)

# The current code
julia> m1(message) = [(indent=0, msg=l) for l in split(chomp(convert(String, string(message))::String), '\n')]
m1 (generic function with 1 method)

# The current PR 
julia> m2(message) = if message isa Base.AnnotatedString
               NamedTuple{(:indent, :msg), Tuple{Int, SubString{<:Union{typeof(message), String}}}}[
                   (indent=0, msg=l) for l in split(chomp(message), '\n')]
           else
               [(indent=0, msg=l) for l in split(
                    chomp(convert(String, string(message))::String), '\n')]
           end
m2 (generic function with 1 method)

# What I'm about to push
julia> m3(message) = if message isa Base.AnnotatedString
               message = Base.AnnotatedString(String(message), Base.annotations(message))
               NamedTuple{(:indent, :msg), Tuple{Int, SubString{<:Union{Base.AnnotatedString{String}, String}}}}[
                   (indent=0, msg=l) for l in split(chomp(message), '\n')]
           else
               [(indent=0, msg=l) for l in split(
                    chomp(convert(String, string(message))::String), '\n')]
           end
m3 (generic function with 1 method)

julia> Base.return_types(m0, (Any,))
1-element Vector{Any}:
 Vector (alias for Array{T, 1} where T)

julia> Base.return_types(m1, (Any,))
1-element Vector{Any}:
 Vector{@NamedTuple{indent::Int64, msg::SubString{String}}} (alias for Array{@NamedTuple{indent::Int64, msg::SubString{String}}, 1})

julia> Base.return_types(m2, (Any,))
1-element Vector{Any}:
 Array{NamedTuple{(:indent, :msg), var"#s178"}, 1} where var"#s178"<:Tuple{Int64, Vararg}

julia> Base.return_types(m3, (Any,))
1-element Vector{Any}:
 Union{Vector{@NamedTuple{indent::Int64, msg::SubString{<:Union{Base.AnnotatedString{String}, String}}}}, Vector{@NamedTuple{indent::Int64, msg::SubString{String}}}}
```

Reply:
(`message` has no type restriction in the signature of `handle_message`)

Reply:
> and the separate codepath is what makes this want for tests.

Added a test for the separate codepath.

Reply:
`handle_message` shifted to being called via `invokelatest` (the "nuclear option" when it comes to eliminating invalidations) in #47889; any changes that occurred prior to that PR may not be required anymore. If you want to be sure but don't feel comfortable using SnoopCompile, just (1) start Julia with `--trace-compile`, (2) define a function that uses logging and run it to force compilation, (3) load InlineStrings, (4) run it again and see if your function got recompiled.